docs(manifest): Improvements to `related_applications` and `prefer_related_applications` (#35906)
mdn/content: files/en-us/web/manifest/related_applications/index.md

{{QuickLinksWithSubpages("/en-US/docs/Web/Manifest")}}{{SeeCompatTable}}

The `related_applications` manifest member is used to specify one or more native applications that are related to your web application.

> > [!NOTE]

Reviewer: This doesn't need to be a note.
```suggestion
```

> > You can set the [`prefer_related_applications`](/en-US/docs/Web/Manifest/prefer_related_applications) manifest member to `true` if you prefer users to install the related native applications over your web app. However, for Chromium-based browsers, `prefer_related_applications` should be set to `false` or omitted.

Reviewer: ```suggestion
Users can be prompted to install related applications if the [`prefer_related_applications`](/en-US/docs/Web/Manifest/prefer_related_applications) manifest member is set to `true` (for Chromium-based browsers, `prefer_related_applications` should be set to `false` or omitted).
```

Author: With or without the note, I think the emphasis here should be on `related_applications`. I've combined this info with the intro para, just like it is in the intro on the `prefer_related_applications` page.

> The `related_applications` manifest member is used to specify one or more native applications that are related to your web application.
It affects user experience when the [`prefer_related_applications`](/en-US/docs/Web/Manifest/prefer_related_applications) manifest member is set to `true`, which indicates a preference for a native application over your web application.

I'll move the part about being prompted to install a related app and caveat for Chromium-based browsers to the "Description" section.

Author: I've now redone the "Description" section to include the note info and also listed the points in a bulleted list for better readability.

## Syntax

```json-nolint
/* Single related native application with both url and id */
"related_applications": [
  {

>     "platform": "play",

Reviewer: Should we comment? I.e.

```suggestion
    "platform": "play", # Google play
```

Author: I haven't added this comment here, just like we're not explaining other details, because those would be covered in "Description" and "Examples".

But this prompted me to make two other changes:
- In the "Syntax" block, I've updated the top-level comments slightly:
   - `/* Single related native application with both url and id */` to 
     `/* Related native application on one platform specified by both url and id */`
   - `/* Single related native application with only id */` to
      `/* Related native application on one platform specfied only by id */`
   - `/* Multiple related native applications */` to
      `/* Related native applications on two platforms */`
- I noticed that the prose in the Examples was generic, so I've made those more specific to call out the platforms.

    "url": "https://play.google.com/store/apps/details?id=com.example.app1",
    "id": "com.example.app1"
  }
]

/* Single related native application with only id */
"related_applications": [
  {
    "platform": "windows",
    "id": "example.app1"
  }
]

/* Multiple related native applications */
"related_applications": [
  {

>     "platform": "play",
>     "url": "https://play.google.com/store/apps/details?id=com.example.app1",
>     "id": "com.example.app1"

Reviewer 2: I'm not sure how you're doing the syntax in manifest in general, but another option is to make it generic in the Syntax section with placeholders like `<app-url>` or `<url>` and describe it in the values section:

```suggestion
    "platform": "<platform>",
    "url": "<url>",
    "id": "<id>"
```

Reviewer 2: More of a comment than a suggestion :)

Reviewer: @bsmth We decided to follow the CSS model for this syntax block and provide concrete examples rather than placeholders (we experimented with all options, including this one in our initial test cases). The description of the value should of course be generic.

Author: It's interesting that you mention that alternative, @bsmth - I had the same idea. But after much brainstorming, we settled on using CSS-style use cases in the "Syntax" section. We feel it is more effective in showing the different possibilities. In some cases, it was getting tricky to convey the true syntax using placeholders.

  },
  {

>     "platform": "itunes",

Reviewer: This platform is not in the list of platforms. What does that mean for our list below?

Author: Well spotted! Changed this to use one from the published list:

```
  {
      "platform": "amazon",
      "url": "https://www.amazon.com/product/dp/B000XA1000"
  }
```

Reviewer: @dipikabh This looks good. I can't see any changes since my last review though. Note this itune example appears in both docs.

Author: Yeah, I am not done addressing all the comments :). I only managed a few before I had to call it a day. The changes are still in my local branch.

Author: Updated/removed the itunes example from both pages

    "url": "https://apps.apple.com/app/example-app1/id123456789"
  }
]
```

### Values

- `related_applications`

  - : An array of objects, each representing a platform-specific native application related to the web app. Each object must include a `platform` property and at least one of either a `url` or an `id` (or both).

    - `platform`

>       - : A string that identifies the platform on which the application can be found. This is a required property. See the [list of available platform values](https://github.com/w3c/manifest/wiki/Platforms).

Reviewer: Should we list a few values?  Just to give a flavour?

```suggestion
      - : A string that identifies the platform on which the application can be found, such as `amazon`, `play`, `windows`.
         [Allowed platform values are listed here](https://github.com/w3c/manifest/wiki/Platforms).
```

Author: Excellent suggestion.
For the link, keeping the link text as "platform values" and avoiding using "here":

> - : A string that identifies the platform on which the application can be found.
>     Examples include `amazon` (Amazon App Store), `play` (Google Play Store), and `windows` (Windows Store).
>     See the complete list of possible [platform values](https://github.com/w3c/manifest/wiki/Platforms).

Reviewer: That would be good

>     - `url`

Reviewer 2: I think you can do an optional inline on both of `url` and `id` and mention "if not defined, an `id` must be provided" on `url`, and the inverse on `id`, what do you reckon?

Reviewer: FWIW I like that idea ^^^

Author: Yes, that's a good idea!

      - : A string that represents the URL at which the platform-specific application can be found.
    - `id`
      - : A string with the ID used to represent the application on the specified platform.

## Description

> A "related application" is an application accessible to the underlying platform that has a relationship with your web app. The `related_applications` manifest member lets you inform users about the platform-specific versions of your web app. These native applications, which are installable by or accessible to the underlying platform, provide functionality similar to your web app and might offer additional features or better integration with user's device. For example, you could specify a native Android application available through the Google Play Store, which offers the same core features as your web app but better integration with the device's notification system.

Reviewer 2: We could make this a little shorter. Is this accurate?

```suggestion
A "related application" is a native application that provides functionality similar to your web app but may offer additional features or better integration with the user's device. For example, you could specify a native Android application available through the Google Play Store, which provides the same core features as your web app but better integration with the device's notification system. The `related_applications` manifest member lets you inform users about the platform-specific versions of your web app.
```

Author: Good one, thanks!

The para can do with some reorganizing and removal of some repeated ideas. However, the Android app example should follow only after we've mentioned `related_applications`, though there is scope for improving the example itself for clarity. How's this:

> A "related application" is a {{Glossary("native")}} application that provides functionality similar to your web app.
It may offer additional features or better integration with users' devices.
The `related_applications` manifest member lets you specify these platform-specific native versions of your web app.
For example, consider you have a native Android app for your product available through the Google Play Store.
It provides the same core features as your web app and integrates better with the device's notification system.
You can use `related_applications` to specify this native Android app in your web app's manifest file.

Author: Slight update here, adding the word "relationship" as it seems to be key in defining "related application". Also introducing a para break after defining "related application":

> A "related application" is a {{Glossary("native")}} application that has a relationship with your web app.
This relationshop means the native app provides functionality similar to your web app, often with additional features or better integration with users' devices.

> The `related_applications` manifest member lets you specify these platform-specific native applications that are related to your web app.
For example, consider you have a native Android app for your product available through the Google Play Store.
It provides the same core features as your web app and integrates better with the device's notification system.
You can use `related_applications` to specify this native Android app in your web app's manifest file.

The `related_applications` member creates a unidirectional relationship: it indicates that the native apps are related to your web app, but does not require the native apps to reference your web app in return. You can specify multiple related applications for different platforms, giving users options across various devices and operating systems.

> The `related_applications` member serves multiple purposes. If you prefer users to install a related native app instead of your web app, you can use it along with the [`prefer_related_applications`](/en-US/docs/Web/Manifest/prefer_related_applications) manifest member. By setting `prefer_related_applications` to `true`, you enable browsers to offer users a choice between installing your web app and the platform-specific version of the native app. Another scenario in which the `related_applications` information might be used is by web crawlers. They might use this data to gather more details about your web app.

Reviewer: "They might use this data to gather more details about your web app."

FWIW I bet no one cares about this - if the preference is not set, this won't be either.

Author: The spec mentions the web crawler point irrespective of the `prefer_related_applications` setting. Hopefully the points described in a list now and their sequence will help clarify this.

(from spec: " Example of usages of the related_applications could be a crawler that would use that information to gather more information about the web application or a browser that could suggest a listed application as an alternative if the user wants to install the web application. ")

{{QuickLinksWithSubpages("/en-US/docs/Web/Manifest")}}{{SeeCompatTable}}

The `related_applications` manifest member is used to specify one or more native applications that are related to your web application.

> It affects user experience when the [`prefer_related_applications`](/en-US/docs/Web/Manifest/prefer_related_applications) manifest member is set to `true`, which indicates a preference for a native application over your web application.

Reviewer: ```suggestion
This can be used with the [`prefer_related_applications`](/en-US/docs/Web/Manifest/prefer_related_applications) manifest member, which indicates a preference for installing either a native application or your web application.
```

  {
    "platform": "play",
    "url": "https://play.google.com/store/apps/details?id=com.example.app1",
    "id": "com.example.app1"
  },
  {
    "platform": "amazon",
    "url": "https://www.amazon.com/product/dp/B000XA1000"
  }
]
```

### Values

- `related_applications`

  - : An array of objects, each representing a platform-specific native application related to the web app. Each object must include a `platform` property and at least one of either a `url` or an `id` (or both).

    - `platform`
      - : A string that identifies the platform on which the application can be found.
        Examples include `amazon` (Amazon App Store), `play` (Google Play Store), and `windows` (Windows Store).
        See the complete list of possible [platform values](https://github.com/w3c/manifest/wiki/Platforms).
    - `url` {{Optional_Inline}}
      - : A string that represents the URL at which the platform-specific application can be found.
        If not specified, an `id` must be provided.
    - `id` {{Optional_Inline}}
      - : A string with the ID used to represent the application on the specified platform.
        If not specified, a `url` must be provided.

## Description

> A "related application" is a {{Glossary("native")}} application that has a relationship with your web app.
> This relationshop means the native app provides functionality similar to your web app, often with additional features or better integration with users' devices.

Reviewer: Let's be succinct
```suggestion
A "related application" is a {{Glossary("native")}} application that provides functionality similar to your web app, often with additional features or better integration with users' devices.
```

> The `related_applications` manifest member lets you specify these platform-specific native applications that are related to your web app.

Reviewer: ```suggestion
The `related_applications` manifest member lets you identify the platform-specific native applications that are related to your web app.
```
